proposals/0035-linalg-matrix.md

```diff
@@ -47,70 +47,816 @@ concept ArithmeticScalar = std::is_arithmetic<T>::value;
+opaque matrix object, and contains an intangible handle that refers to the
+allocated matrix.
+
+The `linalg::Matrix` template type is parameterized based on the matrix
+component data type, dimensions, use, and scope. These parameters restrict where
+and how a matrix can be used.
+
+#### Matrix Use
+
+The `Use` parameter of an instance of a `linalg::Matrix` denotes which argument
+it can be in matrix-matrix operations or matrix-vector operations.
+
+There are three matrix usages: `A`, `B`, and `Accumulator`.
+* The `A` matrix usage denotes a matrix that can be the first argument to binary
+  or ternary algebraic operations.
+* The `B` matrix usage denotes a matrix that can the second argument to binary
+  or ternary algebraic operations.
+* The `Accumulator` matrix usage denotes a matrix that can either be a produced
+  result from a binary arithmetic operation, or the third argument to a ternary
+  algebraic operation.
+
+The matrix use type parameter enables implementations to optimize the storage
+and layout of the matrix prior to tensor operations. It may be expensive on some
+hardware to translate between matrix uses, for that reason we capture the use in
+the type and require explicit conversion in the HLSL API.
+
+Throughout this document a matrix may be described as a matrix of it's use (e.g.
+a matrix with `Use == Accumulator` is an _accumulator matrix_, while a matrix
+with use `A` is an _A matrix_.)
+
+#### Matrix Scope
+
+The `Scope` parameter of an instance of a `linalg::Matrix` denotes the
+uniformity scope of the matrix. The scope impacts which operations can be
+performed on the matrix and may have performance implications depending on the
+implementation.
+
+There are two supported matrix scopes: `Thread` and `Wave`.
+* The `Thread` matrix scope denotes that a matrix's values may vary by thread,
+  which requires that an implementation handle divergent matrix values.
+* The `Wave` matrix scope denotes that a matrix's values are uniform across a
+  wave, which allows an implementation to assume all instances of the matrix
+  across a wave are identical.
+
+Some operations require `Wave` scope matrices, while others can operate on
+`Thread` scope matrices. All operations that can operate on `Thread` scope
+matrices can also operate on `Wave` scope matrices, and there may be significant
+performance benefit when using `Wave` scope matrices.
+
+Throughout this document a matrix may be described as having a scope as
+specified by the `Scope` parameter (e.g. a matrix with `Scope == Thread` is a
+_matrix with thread scope_).
+
+### HLSL API Documentation
+
+#### HLSL Enumerations
+```c++
+enum class MatrixComponentType {
+  Invalid = 0,
+  I1 = 1,
+  I16 = 2,
+  U16 = 3,
+  I32 = 4,
+  U32 = 5,
+  I64 = 6,
+  U64 = 7,
+  F16 = 8,
+  F32 = 9,
+  F64 = 10,
+  SNormF16 = 11,
+  UNormF16 = 12,
+  SNormF32 = 13,
+  UNormF32 = 14,
+  SNormF64 = 15,
+  UNormF64 = 16,
+  PackedS8x32 = 17,

```

Comment:
Something should spell out this type: 4 8bit signed integers, packed into 32 bits.  Little endian when in memory.